### PR TITLE
Audit/final commit fixes

### PR DESCRIPTION
**Pull request by ThomasPiellard**
# Description

Branch based on the final [commit](https://github.com/ThomasPiellard/gnark/tree/audit/final-commit) of the Openzeppelin audit. The branch of the final commit has been kept so it could be referenced in the final report.

This branch contains some fixes after internal review, and also fixes the conflicts with gnark's master branch.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [x] `golangci-lint` does not output errors locally
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by gbotrel — APPROVED** [review]

File: backend/plonk/bn254/solidity.go [hunk 1]
```diff
@@ -18,55 +18,58 @@ const tmplSolidityVerifier = `// SPDX-License-Identifier: Apache-2.0
 
 // Code generated by gnark DO NOT EDIT
 
-pragma solidity {{ .Cfg.PragmaVersion }};
+pragma solidity ^0.8.19;
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:21]: Use from conf

File: backend/plonk/bn254/solidity.go [hunk 2]
```diff
@@ -18,55 +18,58 @@ const tmplSolidityVerifier = `// SPDX-License-Identifier: Apache-2.0
...
 
 contract PlonkVerifier {
 
   uint256 private constant R_MOD = 21888242871839275222246405745257275088548364400416034343698204186575808495617;
   uint256 private constant R_MOD_MINUS_ONE = 21888242871839275222246405745257275088548364400416034343698204186575808495616;
   uint256 private constant P_MOD = 21888242871839275222246405745257275088696311157297823662689037894645226208583;
-  {{ range $index, $element := .Vk.Kzg.G2 }}
+  {{ range $index, $element := .Kzg.G2 }}
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:28]: From Vk (and for the rest of places where using verification key=

File: backend/plonk/bn254/solidity.go [hunk 3]
```diff
@@ -18,55 +18,58 @@ const tmplSolidityVerifier = `// SPDX-License-Identifier: Apache-2.0
...
+  uint256 private constant VK_QO_COM_X = {{ (fpstr .Qo.X) }};
+  uint256 private constant VK_QO_COM_Y = {{ (fpstr .Qo.Y) }};
+  uint256 private constant VK_QK_COM_X = {{ (fpstr .Qk.X) }};
+  uint256 private constant VK_QK_COM_Y = {{ (fpstr .Qk.Y) }};
+  {{ range $index, $element := .S }}
   uint256 private constant VK_S{{ inc $index }}_COM_X = {{ (fpstr $element.X) }};
   uint256 private constant VK_S{{ inc $index }}_COM_Y = {{ (fpstr $element.Y) }};
   {{ end }}
-  uint256 private constant VK_COSET_SHIFT = {{ frstr .Vk.CosetShift }};
+  uint256 private constant VK_COSET_SHIFT = 5;
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:56]: From Vk?

**Review by ivokub — CHANGES_REQUESTED** [review]
Some conflicts when merging other changes. Otherwise looks good.

**Review by ivokub — APPROVED** [review]
All good now! Great work!

**Pull request merged**